psyneulink/core/components/functions/transferfunctions.py

```diff
@@ -425,55 +458,13 @@ def _function(self,
         linear transformation of variable : number or array
 
         """
-        slope = self._get_current_parameter_value(SLOPE, context)
-        intercept = self._get_current_parameter_value(INTERCEPT, context)
-
-        # MODIFIED 11/9/17 NEW:
-        try:
-            # By default, result should be returned as np.ndarray with same dimensionality as input
-            result = variable * slope + intercept
-        except TypeError:
-            if hasattr(variable, "dtype"):
-                # If variable is an array with mixed sizes or types, try item-by-item operation
-                if variable.dtype == object:
-                    result = np.zeros_like(variable)
-                    for i, item in enumerate(variable):
-                        result[i] = variable[i] * slope + intercept
-                else:
-                    raise FunctionError("Unrecognized type for {} of {} ({})".format(VARIABLE, self.name, variable))
-            # KAM 6/28/18: If the variable does not have a "dtype" attr but made it to this line, then it must be of a
-            # type that even np does not recognize -- typically a custom OutputPort variable with items of different
-            # shapes (e.g. variable = [[0.0], [0.0], array([[0.0, 0.0]])] )
-            elif isinstance(variable, list):
-                result = []
-                for variable_item in variable:
-                    result.append(np.multiply(variable_item, slope) + intercept)
-            else:
-                raise FunctionError("Unrecognized type for {} of {} ({})".format(VARIABLE, self.name, variable))
-
-        return self.convert_output_type(result)
-
-    @handle_external_context()
-    def derivative(self, input=None, output=None, context=None):
-        """
-        derivative(input)
+        return variable * slope + intercept
 
-        Derivative of `function <Linear.function>` at **input**.
-
-        Arguments
-        ---------
-
-        input : number
-            value of the input to the Linear transform at which derivative is to be taken.
-
-        Returns
-        -------
-
-        Slope of function :  number or array
-
-        """
-
-        return self._get_current_parameter_value(SLOPE, context)
+    @staticmethod
+    def _derivative(variable=None,
+                    slope=None,
+                    **_):
+        return np.reshape(slope, variable.shape)
```

Comment:
this should use `broadcast` instead of `reshape`. reshape needs compatible shapes (same number of elements)